mmdet3d/models/middle_encoders/voxelsetabstraction.py

```diff
@@ -0,0 +1,290 @@
+        # but I think it doesn't influence trained model accuracy.
+        _, _, y_grid, x_grid = bev_features.shape
+
+        voxel_size_xy = keypoints.new_tensor(self.voxel_size[:2])
+
+        bev_tl_grid_cxy = keypoints.new_tensor(self.point_cloud_range[:2])
+        bev_br_grid_cxy = keypoints.new_tensor(self.point_cloud_range[3:5])
+        bev_tl_grid_cxy.add_(0.5 * voxel_size_xy * scale_factor)
+        bev_br_grid_cxy.sub_(0.5 * voxel_size_xy * scale_factor)
+
+        xy = keypoints[..., :2]
+
+        grid_sample_xy = (xy - bev_tl_grid_cxy[None, None, :]) / (
+            (bev_br_grid_cxy - bev_tl_grid_cxy)[None, None, :])
+
+        grid_sample_xy = (grid_sample_xy * 2 - 1).unsqueeze(1)
+        point_bev_features = F.grid_sample(
+            bev_features, grid=grid_sample_xy, align_corners=True)
+        return point_bev_features.squeeze(2).permute(0, 2, 1).contiguous()
+
+    def aggregate_keypoint_features(self,
+                                    aggregate_module: nn.Module,
+                                    points_xyz: Tensor,
+                                    features: Tensor = None,
+                                    target_xyz: Tensor = None) -> Tuple:
+        """Aggregate keypoint features from one source by aggregate module.
+
+        Args:
+            aggregate_module (nn.Module): A SA module is used to aggregate
+                features.
+            points_xyz (torch.Tensor): (B, N, 3) xyz coordinates of the
+                features.
+            features (torch.Tensor, optional): (B, C, N) features of each
+                point. Default: None.
+            target_xyz (torch.Tensor, optional): (B, M, 3) new coords of
+                the outputs. Default: None.
+
+        Returns:
+            Tensor: (B, M, 3) where M is the number of points.
+                New features xyz.
+            Tensor: (B, M, sum_k(mlps[k][-1])) where M is the number
+                of points. New feature descriptors.
+            Tensor: (B, M) where M is the number of points.
+                Index of the features.
+        """
+        new_xyz, pooled_features, indices = aggregate_module(
+            points_xyz=points_xyz.contiguous(),
+            features=features.contiguous(),
+            target_xyz=target_xyz.contiguous())
+        return new_xyz, pooled_features.transpose(1, 2), indices
+
+    def sample_key_points(self, batch_inputs_dict: dict) -> Tensor:
+        """Sample key points from raw input points.
+
+        Args:
+            batch_inputs_dict (dict): The model input dict which include
+                'points', 'voxel_dict' keys.
+
+                    - points (list[torch.Tensor]): Point cloud of each sample.
+                    - voxel_dict (dict , optional): Voxel dict of each sample.
+
+        Returns:
+            torch.Tensor: Batch key points, shape is (B,N,D),
+                where N is key points num, D is raw points dim.
+        """
+        if self.voxel_center_as_source:
```

Comment:
`voxel_center_as_source` is a little strange, Can we some parameters like `sample_mode` to decide the way of sampling key points. One for `raw_points` and the other for `voxel_center`.

Reply:
Besides, it would be better if we can briefly describe the difference and behaviors in docstring or comments.